### PR TITLE
[SPARK-41295][SPARK-41296][SQL] Rename the error classes

### PR DESCRIPTION
**Pull request by NarekDW**
### What changes were proposed in this pull request?
In the PR, I propose to assign the proper names to the legacy error classes:

- `_LEGACY_ERROR_TEMP_1105` ->  `INVALID_EXTRACT_FIELD_TYPE`
- `_LEGACY_ERROR_TEMP_1106` ->  `INVALID_EXTRACT_BASE_FIELD_TYPE`
- `_LEGACY_ERROR_TEMP_1209` ->  `AMBIGUOUS_REFERENCE_TO_FIELDS` 

Changed error messages for: 
- `_LEGACY_ERROR_TEMP_1106` from `Can't extract value from <child>: need struct type but got <other>.` to `"Can't extract a value from <base>. Need a complex type [struct, array, map] but got <other>."`
- `_LEGACY_ERROR_TEMP_1209` from `"Ambiguous reference to fields <fields>."` to `"Ambiguous reference to the field <field>. It appears <count> times in the schema."`
- `_LEGACY_ERROR_TEMP_1106` from `"Field name should be String Literal, but it's <extraction>."` to `"Field name should be a non-null string literal, but it's <extraction>."`

and modify test suite to use checkError() which checks the error class name, context and etc.

**Also this PR contains an additional change (AMBIGUOUS_REFERENCE_TO_FIELDS) which is not tracked in JIRA.**

### Why are the changes needed?
Proper name improves user experience w/ Spark SQL. fix a bug, you can clarify why it is a bug.

### Does this PR introduce _any_ user-facing change?
Yes, the PR changes an user-facing error message.

### How was this patch tested?
Added additional test cases in `org.apache.spark.sql.errors.QueryCompilationErrorsSuite`

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 1]
```diff
@@ -64,7 +64,7 @@ object ExtractValue {
 
       case (_: ArrayType, _) => GetArrayItem(child, extraction)
 
-      case (MapType(kt, _, _), _) => GetMapValue(child, extraction)
+      case (MapType(_, _, _), _) => GetMapValue(child, extraction)
```
Comment (itholic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:67]: Can I ask what is the intention for this change?
Seems like maybe it is not required to update an error class name??

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala [hunk 2]
```diff
@@ -460,24 +460,44 @@ class ComplexTypeSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+
+    checkError(
+      exception = intercept[AnalysisException](
+        ExtractValue(childStructFiledLiteral, extractionIntegerFieldLiteral, testResolver)),
+      errorClass = "INVALID_EXTRACT_FIELD_TYPE",
+      parameters = Map("extraction" -> extractionIntegerFieldLiteral.toString)
+    )
 
-    checkErrorMessage(structType, IntegerType, "Field name should be String Literal")
-    checkErrorMessage(otherType, StringType, "Can't extract value from")
```
Comment (itholic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:480]: Seems like maybe we can simply update the test by using `checkError` as below?

```diff
-    checkErrorMessage(structType, IntegerType, "Field name should be String Literal")
-    checkErrorMessage(otherType, StringType, "Can't extract value from")
+    checkError(
+      exception = intercept[org.apache.spark.sql.AnalysisException] {
+        ExtractValue(
+          Literal.create(null, structType),
+          Literal.create(null, IntegerType),
+          _ == _)
+      },
+      errorClass = "INVALID_EXTRACT_FIELD_TYPE",
+      parameters = Map("extraction" -> "null")
+    )
+
+    checkError(
+      exception = intercept[org.apache.spark.sql.AnalysisException] {
+        ExtractValue(
+          Literal.create(null, otherType),
+          Literal.create(null, StringType),
+          _ == _)
+      },
+      errorClass = "INVALID_CHILD_FIELD_TYPE",
+      parameters = Map("child" -> "null", "other" -> "string")
+    )
```

Current fix seems a bit harms the readability IMHO.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 1 above)
Comment (NarekDW) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:67]: Are you asking about this particular line of code or the whole change? 
If about the change itself - then at least it improves code readability IMHO

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala (hunk 2 above)
Comment (NarekDW) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:480]: Sure, I've made it more readable, hope it is better now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 1 above)
Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:67]: Oh, yeah I got it it's just code clean-up.
Sounds good

**Comment by itholic** [comment]
cc @srielau @MaxGekk Could you review this error classes when you find some time ??

**Review by srielau — APPROVED** [review]

**Comment by NarekDW** [comment]
@itholic @srielau @MaxGekk anything else is required here from my side?

**Review by itholic — APPROVED** [review]

**Comment by itholic** [comment]
It looks good to me.

Also cc @cloud-fan

**Comment by itholic** [comment]
@NarekDW Can you add `[SPARK-41296]` to PR title instead explaining in PR description ?

`[SPARK-41295][SPARK-41296][SQL] Rename the error classes`

**Comment by NarekDW** [comment]
> 

Sure, done

**Comment by itholic** [comment]
Cool, thanks!!

BTW, if you happen to interested in more contribution to rename error class, could you try resolving SPARK-41302 and SPARK-41488 ??

I believe these are pretty common errors, but haven't been assigned a proper name yet.

**Comment by NarekDW** [comment]
> Cool, thanks!!
> 
> BTW, if you happen to interested in more contribution to rename error class, could you try resolving [SPARK-41302](https://issues.apache.org/jira/browse/SPARK-41302) and [SPARK-41488](https://issues.apache.org/jira/browse/SPARK-41488) ??
> 
> I believe these are pretty common errors, but haven't been assigned a proper name yet. (Please no pressure, I'm just recommending a good item to contribute if you're interested. 😸 )

Yep, I will take a look, no problem, thanks for suggestion :) 
Also I have an off topic question, is it possible to get a JIRA account to be created for me? (I'd like to create couple of tickets there, I wrote a mail to private@spark.apache.org about 2-3 weeks ago, but still didn't get any response...)

**Comment by itholic** [comment]
Oh, I just submit a PR for SPARK-41488, so please take a look SPARK-41302 when you have some time.

**Comment by itholic** [comment]
@srowen Could you happen to help creating JIRA account for @NarekDW when you find some time??

**Comment by NarekDW** [comment]
> @itholic do they want to use the email and name in their Github profile?

Hi @srowen yep, that would be fine.

**Comment by srowen** [comment]
Done

**Comment by itholic** [comment]
Thanks!

**Comment by itholic** [comment]
Gentle ping @MaxGekk as reminder

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala [hunk 3]
```diff
@@ -460,24 +460,38 @@ class ComplexTypeSuite extends SparkFunSuite with ExpressionEvalHelper {
...
-    val otherType = StringType
-
-    def checkErrorMessage(
-      childDataType: DataType,
-      fieldDataType: DataType,
-      errorMessage: String): Unit = {
-      val e = intercept[org.apache.spark.sql.AnalysisException] {
+    checkError(
+      exception = intercept[AnalysisException](
         ExtractValue(
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:465]: Is it possible to reproduce the issue from user space? Using a SQL query, for instance.

Comment (itholic) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:465]: Yeah, if it's not an user-facing we should turn this into `INTERNAL_ERROR`

Comment (NarekDW) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:465]: @MaxGekk I've added 3 test cases in `org.apache.spark.sql.errors.QueryExecutionErrorsSuite` please check

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -738,12 +744,24 @@
     ],
     "sqlState" : "42K05"
   },
+  "INVALID_EXTRACT_BASE_FIELD_TYPE" : {
+    "message" : [
+      "Can't extract value from <base>: need a complex type [struct, array, map] but got <other>."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:749]: nit:
```suggestion
      "Can't extract a value from <base>. Need a complex type [struct, array, map] but got <other>."
```

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -738,12 +744,24 @@
...
+  },
   "INVALID_EXTRACT_FIELD" : {
     "message" : [
       "Cannot extract <field> from <expr>."
     ],
     "sqlState" : "42601"
   },
+  "INVALID_EXTRACT_FIELD_TYPE" : {
+    "message" : [
+      "Field name should be String Literal, but it's <extraction>."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:761]:
```suggestion
      "Field name should be a non-null string literal, but it's <extraction>."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -1117,13 +1117,13 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
     dataType match {
       case StructType(_) =>
         new AnalysisException(
-          errorClass = "_LEGACY_ERROR_TEMP_1105",
+          errorClass = "INVALID_EXTRACT_FIELD_TYPE",
           messageParameters = Map("extraction" -> extraction.toString))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1121]:
```suggestion
          messageParameters = Map("extraction" -> toSQLExpr(extraction)))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 7]
```diff
@@ -1117,13 +1117,13 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
       case other =>
         new AnalysisException(
-          errorClass = "_LEGACY_ERROR_TEMP_1106",
+          errorClass = "INVALID_EXTRACT_BASE_FIELD_TYPE",
           messageParameters = Map(
-            "child" -> child.toString,
+            "base" -> child.toString,
             "other" -> other.catalogString))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1127]:
```suggestion
            "child" -> toSQLExpr(child),
            "other" -> toSQLType(other)))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 8]
```diff
@@ -2104,7 +2104,7 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
 
   def ambiguousReferenceToFieldsError(fields: String): Throwable = {
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1209",
+      errorClass = "AMBIGUOUS_REFERENCE_TO_FIELDS",
       messageParameters = Map("fields" -> fields))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2108]: Fields should be quoted by `toSQLId()`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala:467]: Please, remove duplicate tests here. The added tests are enough.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 8 above)
Comment (NarekDW) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: Done. Also I've changed the error message for this, please check. (added number of field appearance in the schema)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala (hunk 3 above)
Comment (NarekDW) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala]: Done.

**Comment by NarekDW** [comment]
@MaxGekk by the way, don't we need to move these functions from `org.apache.spark.sql.errors.QueryCompilationErrors` to `org.apache.spark.sql.errors.QueryExecutionErrors` ?
From my point of view them are **execution** time errors, not **compilation**...

**Review by MaxGekk — CHANGES_REQUESTED** [review]
All those errors are compilation-wise errors. Please, move tests to `QueryCompilationErrorsSuite`.

**Comment by NarekDW** [comment]
> All those errors are compilation-wise errors. Please, move tests to `QueryCompilationErrorsSuite`.

Done.

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -738,12 +744,24 @@
     ],
     "sqlState" : "42K05"
   },
+  "INVALID_EXTRACT_BASE_FIELD_TYPE" : {
+    "message" : [
+      "Can't extract a value from <base>. Need a complex type [struct, array, map] but got <other>."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:749]: Please, upper case the complex types to be consistent w/ other places, and w/ the function `toSQLType()`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 10]
```diff
@@ -372,13 +372,13 @@ class AnalysisErrorSuite extends AnalysisTest {
   errorTest(
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:372]: Please, migrate on `errorClassTest()`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 11]
```diff
@@ -372,13 +372,13 @@ class AnalysisErrorSuite extends AnalysisTest {
...
     "ambiguous field",
     nestedRelation.select($"top.duplicateField"),
-    "Ambiguous reference to fields" :: "duplicateField" :: Nil,
+    "Ambiguous reference to the field" :: "duplicateField" :: "2" :: Nil,
     caseSensitive = false)
 
   errorTest(
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:378]: Use `errorClassTest()`.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 12]
```diff
@@ -1097,7 +1097,7 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
             nullable = false))))
 
       structLevel2.withColumn("a", $"a".withField("a.b", lit(2)))
-    }.getMessage should include("Ambiguous reference to fields")
+    }.getMessage should include("Ambiguous reference to the field `a`")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1100]: Please, use `checkError`, and don't check the error message. The idea of moving error formats to `error-classes.json` is to allow tech editors modify error messages w/o changing tests.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 13]
```diff
@@ -1650,7 +1650,7 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
     intercept[AnalysisException] {
       sql("SELECT named_struct('a', named_struct('b', 1), 'a', named_struct('c', 2)) struct_col")
         .select($"struct_col".withField("a.c", lit(3)))
-    }.getMessage should include("Ambiguous reference to fields")
+    }.getMessage should include("Ambiguous reference to the field `a`")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1653]: ditto: use `checkError`

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 14]
```diff
@@ -1872,7 +1872,7 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
             nullable = false))))
 
       structLevel2.withColumn("a", $"a".dropFields("a.b"))
-    }.getMessage should include("Ambiguous reference to fields")
+    }.getMessage should include("Ambiguous reference to the field `a`")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1875]: use `checkError`

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 15]
```diff
@@ -2211,7 +2211,7 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
     intercept[AnalysisException] {
       sql("SELECT named_struct('a', named_struct('b', 1), 'a', named_struct('c', 2)) struct_col")
         .select($"struct_col".dropFields("a.c"))
-    }.getMessage should include("Ambiguous reference to fields")
+    }.getMessage should include("Ambiguous reference to the field `a`")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:2214]: Don't check error message, use `checkError` instead of it.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 9 above)
Comment (NarekDW) [reply core/src/main/resources/error/error-classes.json]: Sounds good. Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala (hunk 10 above)
Comment (NarekDW) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala]: Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala (hunk 11 above)
Comment (NarekDW) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala (hunk 12 above)
Comment (NarekDW) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: Yep, done.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala (hunk 13 above)
Comment (NarekDW) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala (hunk 14 above)
Comment (NarekDW) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala (hunk 15 above)
Comment (NarekDW) [reply sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: Done.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.4.
Thank you, @NarekDW and @itholic @srielau @srowen for review.

**Comment by itholic** [comment]
Awesome! Thanks all for working on this 👍

**Pull request closed without merging**